llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -4790,18 +4790,20 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
   // select (trunc nsw X to i1), X, Y --> select (trunc nsw X to i1), -1, Y
   // select (trunc nsw X to i1), Y, X --> select (trunc nsw X to i1), Y, 0
   Value *Trunc;
+  auto Zero = [&]() { return ConstantInt::get(SelType, 0); };
+  auto One = [&]() { return ConstantInt::get(SelType, 1); };
+  auto MinusOne = [&]() { return ConstantInt::getAllOnesValue(SelType); };
   if (match(CondVal, m_NUWTrunc(m_Value(Trunc)))) {
```

Comment:
The simplest way to fix this is to bail out on constant values. `trunc nuw i2 0 to i1` was added into the deferred worklist and never gets constant folded before visiting its users.
```suggestion
  if (match(CondVal, m_NUWTrunc(m_Value(Trunc))) && !isa<Constant>(Trunc)) {
```